Feature/authorize key logic

key-del-jeeinho opened the pull request:
인증키 서비스 내에 인증키 생성 및 조회 기능 구현

key-del-jeeinho committed:
AuthorizeKey 생성 로직 작성
AuthorizeKey 의 Permission 에 대해 분당 요청 제한 횟수를 field 로 가지는 enum 구현

key-del-jeeinho committed:
AuthorizeKeyService 내에 key String 을 통해 AuthorizeKeyDto 를 가져오는 로직 구현
이에 필요한 ApiDto, AuthorizeKeyDto, DtoConverter 등 구현

siwony commented on src/main/java/com/promotion/devconsole/domain/converter/DtoConverter.java:
```diff
@@ -0,0 +1,6 @@
+package com.promotion.devconsole.domain.converter;
+
+public interface DtoConverter<E, D> {
```
이 인터페이스는 어떤 역할을 하나요?

key-del-jeeinho replied:
Dto 와 Entity 간의 converting 이 필요할시 이를 수행하는 Converter 클래스들에서 추출한 인터페이스입니다
E 는 Entity, D 는 Dto 의 약자로
```java
D convertToDto(E entity);
E convertToEntity(D dto);
```
로 두 형식의 객체간의 형변환을 시행합니다

siwony replied:
알겠습니다 :)

siwony commented on src/main/java/com/promotion/devconsole/domain/dto/AuthorizeKeyDto.java:
```diff
@@ -0,0 +1,15 @@
+import com.promotion.devconsole.domain.Permission;
+import lombok.*;
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Getter @Setter
```
DTO에 꼭필요한 경우가 아니라면 객체의 불변성을 위해 `@Setter`, `@NoArgsConstructor` 를 제거해서 불변객체로 만드는게 좋을 거 같아요

`@NoArgsConstructor`는 다음과 같이 해도 좋을 거 같아요`@NoArgsConstructor(access = AccessLevel.PROTECTED)`

`ApiDto` 도 마찬가지 입니다.

siwony replied:
그리고 `@AllArgsConstructor` + `@Builder`로 builder패턴 사용할 수 있어요 :)

key-del-jeeinho replied:
가치있는 피드백들 감사드립니다
**DTO 내 lombok 사용으로 인한 DTO 의 불변성 침해** 에 대한 피드백은 긍정적으로 수용하도록 하겠습니다
혹시 `@AllArgsConstructor` 와 `@Builder` 의 조합으로 builder 패턴을 사용할 수 있다는 건에 대해 예시와 함께 조금더 자세한 설명 가능하신가요??
다시한번 시간내서 피드백해주심에 무한이 감사드립니다~ 😀

siwony replied:
```java
AuthorizeKeyDto authKeyDto =  new AuthorizeKeyDto("authorizeKey" , new ApiDto(), new Permission());
```
위 예제와 같이 객체를 생성하는 것을 다음 예제와 같이 builder 배턴을 이용해 객체를 생성할 수 있어요
```java
AuthorizeKeyDto authKeyDto = AuthorizeKeyDto.builder
                                                 .authorizeKey("authorizeKey")
                                                 .api(new ApiDto())
                                                 .permission(new Permission())
                                                 .build();
```

그냥 참고만 해주세요 :)

key-del-jeeinho replied:
아 `@Builder` 를 사용한 일반적 builder 패턴 말씀하신거였군요 😁
해당 피드백도 긍정적으로 반영해 Constructor 에서 Builder 를 통한 인스턴스 생성 방식으로 변경하도록 하겠습니다.
행복한 저녘시간 되세요 😄😄😄

siwony reviewed (changes requested):
확인했어요 :) 
테스트 코드는 문외한 이여서 잘모르곘네요 ㅜ..

전반적으로 코드가 깔끔해요

siwony reviewed (approved):
모든답변 확인했습니다. 수고하셨어요 :)

key-del-jeeinho committed:
# PR CodeReview 수용
### 모든 Dto 클래스가 불변성을 침해받지 않도록 lombok 어노테이션 제거 / 수정
- `NoArgsConstructor Annotation` 의 `access` field 를 `AccessLevel.PROTECTED` 로 수정
- `Setter Annotation` 제거
### Constructor 를 통해 명확한 의미를 알 수 없는 인스턴스의 생성방식을 Builder 로 변경
- AuthorizeKeyDtoConverter 의 convertToEntity 메서드의 return 인스턴스의 생성방식을 builder 로 변경
### AuthorizeKeyServiceTest 의 authorizeKeyService 초기화 구문의 가독성 증가를 위한 format 변경
